### PR TITLE
DX: Narrow docblock types in Runner and Report

### PR DESCRIPTION
**Pull request by jrmajor**

### CODE REVIEW

File: src/Console/Report/FixReport/ReportSummary.php [hunk 1]
```diff
@@ -21,6 +21,9 @@
  */
 final class ReportSummary
 {
+    /**
+     * @var array<string, array{appliedFixers: list<string>, diff?: string}>
```
Comment (jrmajor) [inline src/Console/Report/FixReport/ReportSummary.php:25]: In `Runner::fixFile()` return type the `diff` key is not optional, but there are tests for `ReportSummary` with no diff, and some reporters have `empty($fixResult['diff'])` check. Can diff be empty under some circumstances, or is it a residue from earlier versions?

File: dev-tools/composer.json [hunk 2]
```diff
@@ -10,8 +10,8 @@
         "maglnet/composer-require-checker": "3.3.0",
         "mi-schi/phpmd-extension": "^4.3",
         "phpmd/phpmd": "^2.10.2",
-        "phpstan/phpstan": "^1.2",
-        "phpstan/phpstan-phpunit": "^1.0"
+        "phpstan/phpstan": "^1.8",
```
[the remark this answers is not in the document]
Comment (jrmajor) [reply dev-tools/composer.json]: Rebased onto #6467 and removed this commit.

File: src/Console/Report/FixReport/ReportSummary.php (hunk 1 above)
Comment (keradus) [reply src/Console/Report/FixReport/ReportSummary.php:25]: the code-diff generation is optional. with that, you may not assume that it's always present.

what does it mean for types? I'm not sure if we return `null` or `empty string` when we do not generate diff

File: src/Console/Report/FixReport/ReporterFactory.php [hunk 3]
```diff
@@ -25,19 +24,18 @@
...
     /**
-     * @var ReporterInterface[]
+     * @var array<string, ReporterInterface>
      */
     private array $reporters = [];
 
     public function registerBuiltInReporters(): self
     {
-        /** @var null|string[] $builtInReporters */
+        /** @var null|list<string> $builtInReporters */
```
Comment (keradus) [inline src/Console/Report/FixReport/ReporterFactory.php:33]: why?

File: tests/Runner/FileFilterIteratorTest.php [hunk 4]
```diff
@@ -172,7 +172,7 @@ public function testWithoutDispatcher(): void
     public function testInvalidIterator(): void
     {
         $filter = new FileFilterIterator(
-            new \ArrayIterator([__FILE__]),
+            new \ArrayIterator([__FILE__]), // @phpstan-ignore-line
```
Comment (keradus) [inline tests/Runner/FileFilterIteratorTest.php:175]: why?

File: src/Console/Report/FixReport/ReporterFactory.php (hunk 3 above)
Comment (jrmajor) [reply src/Console/Report/FixReport/ReporterFactory.php:33]: You have already started that discussion in #6293, please see https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/6293#discussion_r820157214 and let me know whether that convinces you or do you still want me to revert these changes :)

File: tests/Runner/FileFilterIteratorTest.php (hunk 4 above)
Comment (jrmajor) [reply tests/Runner/FileFilterIteratorTest.php:175]: I've added `\Traversable<\SplFileInfo>` type hint to `FileFilterIterator::__construct()`, so PHPStan now detects invalid usage. Probably no need to care about that anyway, as @julienfalque made the same changes in #6468 and just deleted that test.

File: src/Console/Report/FixReport/ReportSummary.php (hunk 1 above)
Comment (jrmajor) [reply src/Console/Report/FixReport/ReportSummary.php:25]: We have a `NullDiffer`, which returns an empty string, so probably that's the case.

I'll update optional types and add an empty string to the tests where it's missing.

**Review by julienfalque — APPROVED** [review]

**Comment by julienfalque** [comment]
Thank you @jrmajor.

**Pull request merged**